File: App/BNKScopeField/Model/ClusterStore.swift
import Foundation
import Observation
import BNKKit

/// A cluster as the app knows it: the kubeconfig context, plus whatever probing
/// it has learned.
@Observable
@MainActor
final class ManagedCluster: Identifiable {
    let context: Kubeconfig.Context
    let sourceFile: String

    /// `nonisolated` because the identity is the context name, which is a `let`
    /// fixed at init — SwiftUI reads it off the main actor while diffing lists,
    /// and nothing about it can race.
    nonisolated var id: String { context.name }

    /// What to call this cluster in the UI.
    ///
    /// Context names are written for kubectl, not for reading:
    /// `kubernetes-admin@dpu-cplane-tenant1` says one useful word and eleven
    /// that are the same on every context. The useful word is the cluster, so
    /// that is what is shown — unless the cluster is called something kubeadm
    /// picked, in which case its address is the one thing that tells it apart.
    ///
    /// The file it came from was tried and is worse: after import every cluster
    /// owns a file named after its context, so a kubeadm cluster came out as
    /// `kubernetes-admin-kubernetes.kubeconfig`, which is the same eleven words
    /// again with the file's suffix on the end. Two such clusters on one lab
    /// read as one name twice; as `host:port` they read as two.
    nonisolated var displayName: String {
        let generic: Set<String> = ["kubernetes", "default", "cluster.local", "kind"]
        let afterAt = context.name.split(separator: "@").last.map(String.init)
        for candidate in [afterAt, context.clusterName].compactMap({ $0 }) {
            if !generic.contains(candidate) { return candidate }
        }
        if let host = context.server.host() {
            return context.server.port.map { "\(host):\($0)" } ?? host
        }
        return context.name
    }

    enum Reach: Equatable {
        case unprobed
        case reachable(version: String, nodes: Int, ready: Int)
        case unreachable(String)
        case unusable(String)
    }

    var reach: Reach = .unprobed
    var roles: Set<Role> = []
    var tmmPods: [K8s.Pod] = []
    /// What k0rdent this cluster is, if any — edition, version, and the object
    /// it is managed as. Empty on a cluster k0rdent has never touched.
    var k0rdent = K0rdent.Fingerprint()
    /// The API group versions discovery reported, so a screen can address an
    /// operator's API without guessing which version it serves.
    var apiGroups: [String: String] = [:]
    /// GPU extended resources offered across the nodes, already summed —
    /// `["GA104GL_RTX_A4000 ×2"]`.
    var gpuDevices: [String] = []
    /// Bumped every time probing finishes.
    ///
    /// A screen that loads from what probing discovered cannot key its work on
    /// the cluster's identity alone: the first render happens before the probe
    /// answers, and selecting the same cluster again afterwards is not a change
    /// SwiftUI will notice. Keying on this makes "the facts changed" a distinct
    /// event from "the selection changed".
    private(set) var probeGeneration = 0

    /// What a cluster has on it, as the sidebar badges it.
    ///
    /// Not a taxonomy — a cluster is frequently several of these at once, and
    /// the useful reading is the combination. A `.managed` cluster that is also
    /// `.gpu` and `.kubevirt` is the one running tenant VMs on passed-through
    /// cards, which is the thing worth finding in a list.
    enum Role: String, CaseIterable, Comparable {
        case bnk = "BNK", dpu = "DPU", nico = "NICo"
        /// Serves the k0rdent API: this is a management cluster.
        case k0rdent = "k0rdent"
        /// Provisioned or adopted by a k0rdent management cluster elsewhere.
        case managed = "k0rdent-managed"
        /// At least one node advertises a GPU as an extended resource.
        case gpu = "GPU"
        case kubevirt = "KubeVirt"

        static func < (a: Role, b: Role) -> Bool { a.rawValue < b.rawValue }
    }

    private var cached: KubeClient?
    private var probeInFlight: Task<Void, Never>?
    private var probeSerial = 0

    init(context: Kubeconfig.Context, sourceFile: String) {
        self.context = context
        self.sourceFile = sourceFile
        if case .unsupported(let why) = context.auth { reach = .unusable(why) }
    }

    func client() throws -> KubeClient {
        if let cached { return cached }
        let c = try KubeClient(context: context, tag: "bnkscope.field.\(context.name)")
        cached = c
        return c
    }

    var isUsable: Bool {
        if case .unusable = reach { return false }
        return true
    }

    /// Ask the cluster what it is.
    ///
    /// TMM, DPU and NICo are read from pod labels rather than namespace names,
    /// because on a real deployment these live on different clusters and the
    /// namespaces vary by install shape — the same reason bnkscope detects them
    /// this way. k0rdent and KubeVirt are read from API discovery instead:
    /// both are operators, an operator's namespace is a deployment choice, and
    /// the API it serves is not.
    /// Probe, or join the probe already running.
    ///
    /// Two callers ask for this concurrently — the store probes everything at
    /// launch while a screen probes the cluster it needs — and two probes
    /// interleaving on one cluster left it reporting "no route from this iPad"
    /// while its data sat on the screen beside the message.
    func probe() async {
        if let probeInFlight {
            await probeInFlight.value
            return
        }
        await runProbe()
    }

    /// Probe for a caller that has just changed the cluster.
    ///
    /// Joining an in-flight probe is wrong here. That probe listed the pods
    /// before the change landed, so adopting its answer shows the cluster as it
    /// was a moment ago: an exporter this app just installed reads as still
    /// missing, the overview keeps warning about it, and nothing starts
    /// scraping the pod it was added to.
    func probeReflectingChange() async {
        if let probeInFlight { await probeInFlight.value }
        await runProbe()
    }

    private func runProbe() async {
        probeSerial += 1
        let mine = probeSerial
        let task = Task { await performProbe() }
        probeInFlight = task
        await task.value
        // Only retire the handle if it is still ours — a caller that waited on
        // us has since installed its own, and clearing that one would let a
        // third caller start a duplicate probe.
        if probeSerial == mine { probeInFlight = nil }
    }

    private func performProbe() async {
        guard isUsable else { return }
        do {
            let c = try client()
            let version = try await c.version()
            let nodes = try await c.nodes()
            reach = .reachable(version: version.gitVersion,
                               nodes: nodes.count,
                               ready: nodes.filter(\.isReady).count)

            var found: Set<Role> = []
            let tmm = try await c.pods(labelSelector: "app=f5-tmm")
            if !tmm.isEmpty { found.insert(.bnk) }
            tmmPods = tmm
            // A svc.dpu.nvidia.com/ label means the workload is wired through
            // the DPU service API. It does NOT mean the DPF operator is here —
            // that is a different API group, and on this lab it is installed on
            // neither reachable cluster. The badge used to say DPF and was
            // pointing at something true under the wrong name.
            if tmm.contains(where: { ($0.metadata.labels ?? [:]).keys.contains { $0.hasPrefix("svc.dpu.nvidia.com/") } }) {
                found.insert(.dpu)
            }
            if try await !c.pods(labelSelector: "app.kubernetes.io/name=nico-api").isEmpty {
                found.insert(.nico)
            }

            // One discovery call answers "is k0rdent here" and "is KubeVirt
            // here" together, and hands both the version each serves. Asked
            // before either detector runs so neither has to hard-code a path.
            apiGroups = (try? await c.apiGroups()) ?? [:]

            k0rdent = await c.k0rdentFingerprint(groups: apiGroups)
            switch k0rdent.role {
            case .management: found.insert(.k0rdent)
            case .managed:    found.insert(.managed)
            case nil:         break
            }

            if c.kubeVirtVersion(groups: apiGroups) != nil { found.insert(.kubevirt) }

            // Summed across nodes rather than reported per node: what decides
            // whether a VM can be placed is how many cards the cluster has, and
            // a two-node cluster with one card each reads the same as one node
            // with two until you try to give a single VM both.
            var gpus: [String: Int] = [:]
            for node in nodes {
                for gpu in node.gpuResources { gpus[gpu.name, default: 0] += gpu.count }
            }
            gpuDevices = gpus.map { $0.value > 1 ? "\($0.key) ×\($0.value)" : $0.key }.sorted()
            if !gpus.isEmpty { found.insert(.gpu) }

            roles = found
        } catch {
            reach = .unreachable(Self.explain(error))
            tmmPods = []
            k0rdent = K0rdent.Fingerprint()
            apiGroups = [:]
            gpuDevices = []
        }
        probeGeneration += 1
    }

    /// A URLError says almost nothing useful on its own. Three cases actually
    /// happen here, and each has a different fix: a kubeconfig pointing
    /// somewhere this device cannot route, a certificate the device will not
    /// accept, and the OS refusing this app the local network outright.
    static func explain(_ error: Error) -> String {
        guard let url = error as? URLError else { return String(describing: error) }
        switch url.code {
        case .cannotConnectToHost, .timedOut, .networkConnectionLost:
            return "no route to this address from \(Self.thisDevice)"
        case .notConnectedToInternet:
            // A Local Network denial arrives under the same code as a device
            // with no network at all. Every cluster this app is pointed at is
            // on a local subnet, so when the address is one only the local
            // network can reach, the permission is much the likelier reading —
            // and unlike "offline", it names its own fix.
            if let host = url.failingURL?.host(), Net.isLocal(host: host) {
                return "\(Self.thisDevice) is refusing this app the local network — "
                     + "switch bnkscope Field on under \(Self.localNetworkSetting)"
            }
            return "\(Self.thisDevice) is not on a network"
        case .secureConnectionFailed, .serverCertificateUntrusted,
             .serverCertificateHasBadDate, .serverCertificateNotYetValid:
            return "the TLS handshake failed — check the certificate authority in the kubeconfig"
        default:
            return url.localizedDescription
        }
    }

    // What to call the machine, and where its privacy switch lives. The app
    // runs on both, and an iPad's wording on a Mac reads as a bug.
    #if os(macOS)
    static let thisDevice = "this Mac"
    static let localNetworkSetting = "System Settings › Privacy & Security › Local Network"
    static let importSource = "from Finder"
    #else
    static let thisDevice = "this iPad"
    static let localNetworkSetting = "Settings › Privacy & Security › Local Network"
    static let importSource = "from Files"
    #endif
}

/// Every cluster the app holds, and the kubeconfigs they came from.
@Observable
@MainActor
final class ClusterStore {
    private(set) var clusters: [ManagedCluster] = []
    private(set) var files: [String] = []
    var selected: ManagedCluster.ID?
    var importError: String?

    /// Kubeconfigs live in Application Support, excluded from backup. The
    /// certificates inside them are also imported into the keychain, which is
    /// what actually gets used; this copy is what lets the app rebuild that
    /// after a reinstall without asking for the file again.
    private static var directory: URL {
        let base = URL.applicationSupportDirectory.appending(path: "kubeconfigs")
        try? FileManager.default.createDirectory(at: base, withIntermediateDirectories: true)
        return base
    }

    func load() {
        clusters.removeAll()
        files.removeAll()
        let urls = (try? FileManager.default.contentsOfDirectory(at: Self.directory,
                                                                 includingPropertiesForKeys: nil)) ?? []
        var files = urls.sorted { $0.lastPathComponent < $1.lastPathComponent }
        if files.contains(where: { migrateMultiContext($0) }) {
            files = ((try? FileManager.default.contentsOfDirectory(
                at: Self.directory, includingPropertiesForKeys: nil)) ?? [])
                .sorted { $0.lastPathComponent < $1.lastPathComponent }
        }
        for url in files { adopt(url) }
        selectSomethingUseful()
    }

    /// Prefer a cluster that has something to show. Landing on a reachable
    /// cluster with no TMM pods and the message "looking for TMM pods" reads as
    /// a fault when it is only a bad default.
    private func selectSomethingUseful() {
        guard selected == nil || current?.isUsable != true else { return }
        selected = clusters.first { $0.roles.contains(.bnk) }?.id
            ?? clusters.first { if case .reachable = $0.reach { return true } else { return false } }?.id
            ?? clusters.first(where: \.isUsable)?.id
    }

    /// Copy an imported file in and adopt its contexts.
    func importKubeconfig(from source: URL, named name: String? = nil) {
        importError = nil
        let needsScope = source.startAccessingSecurityScopedResource()
        defer { if needsScope { source.stopAccessingSecurityScopedResource() } }
        do {
            let data = try Data(contentsOf: source)
            // Parse before storing: a file that is not a kubeconfig should be
            // refused at the picker, not discovered on the next launch.
            _ = try Kubeconfig(yaml: String(decoding: data, as: UTF8.self))
            // Split on the way in: after import a cluster is its own thing, and
            // which file it arrived in stops mattering.
            for (name, text) in try Kubeconfig.split(yaml: String(decoding: data, as: UTF8.self)) {
                let target = Self.directory.appending(path: Self.filename(for: name))
                try text.write(to: target, atomically: true, encoding: .utf8)
                try? FileManager.default.setAttributes(
                    [.protectionKey: FileProtectionType.complete], ofItemAtPath: target.path)
                adopt(target)
            }
            selectSomethingUseful()
        } catch {
            importError = String(describing: error)
        }
    }

    /// Forget one cluster.
    ///
    /// Each cluster owns its own file, so this takes that file and nothing
    /// else. Removal used to be per FILE, which on a kubeconfig holding three
    /// contexts meant removing one dead cluster took the two working ones with
    /// it.
    func remove(_ cluster: ManagedCluster) {
        if case .clientCertificate(let certPEM, _) = cluster.context.auth {
            Identity.forget(tag: "bnkscope.field.\(cluster.context.name)", certPEM: certPEM)
        }
        try? FileManager.default.removeItem(at: Self.directory.appending(path: cluster.sourceFile))
        clusters.removeAll { $0.id == cluster.id }
        files.removeAll { $0 == cluster.sourceFile }
        if let selected, !clusters.contains(where: { $0.id == selected }) {
            self.selected = nil
            selectSomethingUseful()
        }
    }

    /// The other clusters that came out of the same file. They stay.
    func siblings(of cluster: ManagedCluster) -> [String] {
        clusters.filter { $0.sourceFile == cluster.sourceFile && $0.id != cluster.id }
            .map(\.displayName)
    }

    func probeAll() async {
        await withTaskGroup(of: Void.self) { group in
            for cluster in clusters where cluster.isUsable {
                group.addTask { await cluster.probe() }
            }
        }
        // Roles are only known once probing has answered, so the default choice
        // is worth revisiting now rather than at load.
        selected = nil
        selectSomethingUseful()
    }

    var current: ManagedCluster? {
        clusters.first { $0.id == selected }
    }

    /// A file name that is one cluster's, and safe on disk.
    static func filename(for context: String) -> String {
        let safe = context.map { $0.isLetter || $0.isNumber || $0 == "-" || $0 == "." ? $0 : "-" }
        return String(safe) + ".kubeconfig"
    }

    /// Anything imported before the split is still one file holding several
    /// clusters. Splitting it on load means an old install behaves like a new
    /// one rather than keeping the coupling for ever.
    private func migrateMultiContext(_ url: URL) -> Bool {
        guard let text = try? String(contentsOf: url, encoding: .utf8),
              let parts = try? Kubeconfig.split(yaml: text), parts.count > 1 else { return false }
        for (name, part) in parts {
            let target = Self.directory.appending(path: Self.filename(for: name))
            try? part.write(to: target, atomically: true, encoding: .utf8)
        }
        try? FileManager.default.removeItem(at: url)
        return true
    }

    private func adopt(_ url: URL) {
        guard let text = try? String(contentsOf: url, encoding: .utf8),
              let config = try? Kubeconfig(yaml: text) else { return }
        files.append(url.lastPathComponent)
        for context in config.contexts where !clusters.contains(where: { $0.id == context.name }) {
            clusters.append(ManagedCluster(context: context, sourceFile: url.lastPathComponent))
        }
    }
}

File: dart/bnk_engines/lib/src/cluster_store.dart
import 'dart:async';
import 'dart:io';

import 'package:bnk_kit/bnk_kit.dart';

import 'observable.dart';

/// The words that differ by machine: what to call it, where its privacy
/// switch lives, where a file comes from. Decided by the app, never here.
class DeviceWords {
  final String thisDevice;
  final String localNetworkSetting;
  final String importSource;
  const DeviceWords({required this.thisDevice, required this.localNetworkSetting, required this.importSource});

  static const generic = DeviceWords(
      thisDevice: 'this device',
      localNetworkSetting: 'the system privacy settings',
      importSource: 'from a file');
}

/// Whether the apiserver answered, and what it said about itself.
sealed class Reach {
  const Reach();
}

class Unprobed extends Reach {
  const Unprobed();
}

class Reachable extends Reach {
  final String version;
  final int nodes;
  final int ready;
  const Reachable({required this.version, required this.nodes, required this.ready});
}

class Unreachable extends Reach {
  final String why;
  const Unreachable(this.why);
}

class Unusable extends Reach {
  final String why;
  const Unusable(this.why);
}

/// What a cluster has on it, as the sidebar badges it.
///
/// Not a taxonomy: a cluster is frequently several of these at once, and the
/// useful reading is the combination. A managed cluster that is also gpu and
/// kubevirt is the one running tenant VMs on passed-through cards, which is
/// the thing worth finding in a list.
enum ClusterRole {
  bnk('BNK'),
  dpu('DPU'),
  nico('NICo'),

  /// Serves the k0rdent API: this is a management cluster.
  k0rdent('k0rdent'),

  /// Provisioned or adopted by a k0rdent management cluster elsewhere.
  managed('k0rdent-managed'),

  /// At least one node advertises a GPU as an extended resource.
  gpu('GPU'),
  kubevirt('KubeVirt');

  final String label;
  const ClusterRole(this.label);
}

/// A cluster as the app knows it: the kubeconfig context, plus whatever
/// probing it has learned.
class ManagedCluster extends Observable {
  final KubeContext context;
  final String sourceFile;
  final DeviceWords words;

  Reach _reach = const Unprobed();
  Set<ClusterRole> _roles = {};
  List<Pod> _tmmPods = const [];
  K0rdentFingerprint _k0rdent = K0rdentFingerprint();
  Map<String, String> _apiGroups = const {};
  List<String> _gpuDevices = const [];
  int _probeGeneration = 0;

  KubeClient? _cached;
  Future<void>? _probeInFlight;
  int _probeSerial = 0;

  ManagedCluster({required this.context, required this.sourceFile, this.words = DeviceWords.generic}) {
    if (context.auth case UnsupportedAuth(:final reason)) _reach = Unusable(reason);
  }

  String get id => context.name;

  Reach get reach => _reach;
  Set<ClusterRole> get roles => _roles;
  List<Pod> get tmmPods => _tmmPods;

  /// What k0rdent this cluster is, if any. Empty on a cluster k0rdent has
  /// never touched.
  K0rdentFingerprint get k0rdent => _k0rdent;

  /// The API group versions discovery reported, so a screen can address an
  /// operator's API without guessing which version it serves.
  Map<String, String> get apiGroups => _apiGroups;

  /// GPU extended resources offered across the nodes, already summed:
  /// `["GA104GL_RTX_A4000 ×2"]`.
  List<String> get gpuDevices => _gpuDevices;

  /// Bumped every time probing finishes, so a screen can tell "the facts
  /// changed" from "the selection changed".
  int get probeGeneration => _probeGeneration;

  /// What to call this cluster in the UI.
  ///
  /// Context names are written for kubectl, not for reading:
  /// `kubernetes-admin@dpu-cplane-tenant1` says one useful word and eleven
  /// that are the same on every context. The useful word is the cluster, so
  /// that is what is shown, unless the cluster is called something kubeadm
  /// picked, in which case its address is the one thing that tells it apart.
  String get displayName {
    const generic = {'kubernetes', 'default', 'cluster.local', 'kind'};
    final afterAt = context.name.split('@').last;
    for (final candidate in [afterAt, context.clusterName]) {
      if (!generic.contains(candidate)) return candidate;
    }
    final server = context.server;
    if (server.host.isNotEmpty) {
      return server.hasPort ? '${server.host}:${server.port}' : server.host;
    }
    return context.name;
  }

  bool get isUsable => _reach is! Unusable;

  KubeClient client() => _cached ??= KubeClient(context);

  /// The client, or null when one cannot be built: the certificate did not
  /// parse, or the context needs a binary.
  KubeClient? get clientOrNull {
    if (!isUsable) return null;
    try {
      return client();
    } catch (_) {
      return null;
    }
  }

  /// Probe, or join the probe already running.
  ///
  /// Two callers ask for this concurrently, the store probes everything at
  /// launch while a screen probes the cluster it needs, and two probes
  /// interleaving on one cluster left it reporting "no route" while its data
  /// sat on the screen beside the message.
  Future<void> probe() async {
    final running = _probeInFlight;
    if (running != null) return running;
    await _runProbe();
  }

  /// Probe for a caller that has just changed the cluster.
  ///
  /// Joining an in-flight probe is wrong here. That probe listed the pods
  /// before the change landed, so adopting its answer shows the cluster as
  /// it was a moment ago.
  Future<void> probeReflectingChange() async {
    final running = _probeInFlight;
    if (running != null) await running;
    await _runProbe();
  }

  /// Let the connection pool go with the cluster: an idle keep-alive is a
  /// timer nobody will use.
  @override
  void dispose() {
    _cached?.close();
    _cached = null;
    super.dispose();
  }

  Future<void> _runProbe() async {
    _probeSerial++;
    final mine = _probeSerial;
    final task = _performProbe();
    _probeInFlight = task;
    await task;
    // Only retire the handle if it is still ours: a caller that waited on us
    // has since installed its own, and clearing that one would let a third
    // caller start a duplicate probe.
    if (_probeSerial == mine) _probeInFlight = null;
  }

  /// Ask the cluster what it is.
  ///
  /// TMM, DPU and NICo are read from pod labels rather than namespace names,
  /// because on a real deployment these live on different clusters and the
  /// namespaces vary by install shape. k0rdent and KubeVirt are read from
  /// API discovery instead: both are operators, an operator's namespace is a
  /// deployment choice, and the API it serves is not.
  Future<void> _performProbe() async {
    if (!isUsable) return;
    try {
      final c = client();
      final version = await c.version();
      final nodes = await c.nodes();
      _reach = Reachable(
          version: version.gitVersion, nodes: nodes.length, ready: nodes.where((n) => n.isReady).length);
      notify();

      final found = <ClusterRole>{};
      final tmm = await c.pods(labelSelector: 'app=f5-tmm');
      if (tmm.isNotEmpty) found.add(ClusterRole.bnk);
      _tmmPods = tmm;
      // A svc.dpu.nvidia.com/ label means the workload is wired through the
      // DPU service API. It does NOT mean the DPF operator is here; that is
      // a different API group.
      if (tmm.any((p) => (p.metadata.labels ?? const {}).keys.any((k) => k.startsWith('svc.dpu.nvidia.com/')))) {
        found.add(ClusterRole.dpu);
      }
      if ((await c.pods(labelSelector: 'app.kubernetes.io/name=nico-api')).isNotEmpty) {
        found.add(ClusterRole.nico);
      }

      // One discovery call answers "is k0rdent here" and "is KubeVirt here"
      // together. Not swallowed: a discovery call that fails used to read as
      // an empty map, and an empty map is a claim. A probe that cannot
      // finish has not answered.
      _apiGroups = await c.apiGroups();

      _k0rdent = await c.k0rdentFingerprint(_apiGroups);
      switch (_k0rdent.role) {
        case K0rdentRole.management:
          found.add(ClusterRole.k0rdent);
        case K0rdentRole.managed:
          found.add(ClusterRole.managed);
        case null:
          break;
      }
      if (c.kubeVirtVersion(_apiGroups) != null) found.add(ClusterRole.kubevirt);

      // Summed across nodes rather than reported per node: what decides
      // whether a VM can be placed is how many cards the cluster has.
      final gpus = <String, int>{};
      for (final node in nodes) {
        for (final gpu in node.gpuResources) {
          gpus[gpu.name] = (gpus[gpu.name] ?? 0) + gpu.count;
        }
      }
      _gpuDevices = [for (final e in gpus.entries) e.value > 1 ? '${e.key} ×${e.value}' : e.key]..sort();
      if (gpus.isNotEmpty) found.add(ClusterRole.gpu);

      _roles = found;
    } catch (e) {
      _reach = Unreachable(explain(e, words: words));
      // Everything probing derived stays as last known, together. The reach
      // says the cluster cannot be read now; the roles, the fingerprint and
      // the groups say what it was when it could.
      _tmmPods = const [];
    }
    _probeGeneration++;
    notify();
  }

  /// A socket error says almost nothing useful on its own. Three cases
  /// actually happen here, and each has a different fix: a kubeconfig
  /// pointing somewhere this device cannot route, a certificate the device
  /// will not accept, and the OS refusing this app the local network.
  static String explain(Object error, {DeviceWords words = DeviceWords.generic}) {
    if (error is UnusableFailure) return error.why;
    if (error is HandshakeException || error is CertificateException || error is TlsException) {
      return 'the TLS handshake failed — check the certificate authority in the kubeconfig';
    }
    if (error is TimeoutException) return 'no route to this address from ${words.thisDevice}';
    if (error is SocketException) {
      final host = error.address?.host ?? '';
      // A Local Network denial on iOS and macOS arrives as the OS refusing
      // the connection outright. Every cluster this app is pointed at is on
      // a local subnet, so when the address is one only the local network
      // can reach and the OS said "not permitted", the permission is much
      // the likelier reading, and unlike "no route" it names its own fix.
      final message = error.osError?.message.toLowerCase() ?? '';
      if (host.isNotEmpty && Net.isLocal(host) && message.contains('not permitted')) {
        return '${words.thisDevice} is refusing this app the local network — '
            'switch bnkscope Field on under ${words.localNetworkSetting}';
      }
      // macOS answers a Local Network denial with "No route to host", the
      // same words as a genuinely unroutable address. Say both, because
      // the reader can only fix one of them from here.
      if (host.isNotEmpty && Net.isLocal(host) && message.contains('no route to host')) {
        return 'no route to this local address from ${words.thisDevice}. If it is on the same network, '
            '${words.thisDevice} may be refusing this app the local network: '
            'switch bnkscope Field on under ${words.localNetworkSetting}';
      }
      return 'no route to this address from ${words.thisDevice}';
    }
    return error.toString();
  }
}

/// Every cluster the app holds, and the kubeconfigs they came from.
///
/// Kubeconfigs live in [directory], one file per context; the app decides
/// where that is (application support, excluded from backup).
class ClusterStore extends Observable {
  final Directory directory;
  final DeviceWords words;

  List<ManagedCluster> _clusters = [];
  List<String> _files = [];
  String? _selected;
  String? _importError;

  /// Whether the selection was this store's guess rather than the user's
  /// pick. A guess is revisited once probing has answered; a pick is not.
  bool _selectionWasAutomatic = true;

  ClusterStore({required this.directory, this.words = DeviceWords.generic});

  List<ManagedCluster> get clusters => _clusters;
  List<String> get files => _files;
  String? get importError => _importError;

  String? get selected => _selected;
  set selected(String? id) {
    _selected = id;
    _selectionWasAutomatic = false;
    notify();
  }

  ManagedCluster? get current {
    for (final c in _clusters) {
      if (c.id == _selected) return c;
    }
    return null;
  }

  Future<void> load() async {
    for (final c in _clusters) {
      c.dispose();
    }
    _clusters = [];
    _files = [];
    await directory.create(recursive: true);
    var files = await _listFiles();
    var migrated = false;
    for (final f in files) {
      if (await _migrateMultiContext(f)) migrated = true;
    }
    if (migrated) files = await _listFiles();
    for (final f in files) {
      await _adopt(f);
    }
    _selectSomethingUseful();
    notify();
  }

  Future<List<File>> _listFiles() async {
    final out = <File>[];
    await for (final e in directory.list()) {
      if (e is File) out.add(e);
    }
    out.sort((a, b) => _basename(a).compareTo(_basename(b)));
    return out;
  }

  static String _basename(File f) => f.uri.pathSegments.last;

  /// Prefer a cluster that has something to show. Landing on a reachable
  /// cluster with no TMM pods reads as a fault when it is only a bad default.
  void _selectSomethingUseful() {
    if (_selected != null && current?.isUsable == true) return;
    ManagedCluster? pick;
    for (final c in _clusters) {
      if (c.roles.contains(ClusterRole.bnk)) {
        pick = c;
        break;
      }
    }
    pick ??= _clusters.where((c) => c.reach is Reachable).firstOrNull;
    pick ??= _clusters.where((c) => c.isUsable).firstOrNull;
    _selected = pick?.id;
    _selectionWasAutomatic = true;
  }

  /// Adopt the contexts of a kubeconfig's text, one file per context.
  ///
  /// Parsed before storing: a file that is not a kubeconfig should be
  /// refused at the picker, not discovered on the next launch. Split on the
  /// way in: after import a cluster is its own thing, and which file it
  /// arrived in stops mattering.
  Future<void> importKubeconfig(String text) async {
    _importError = null;
    try {
      Kubeconfig.parse(text);
      await directory.create(recursive: true);
      for (final part in Kubeconfig.split(text)) {
        final target = File('${directory.path}/${filenameFor(part.name)}');
        await target.writeAsString(part.yaml, flush: true);
        await _adopt(target);
      }
      _selectSomethingUseful();
    } catch (e) {
      _importError = e.toString();
    }
    notify();
  }

  /// Forget one cluster. Each cluster owns its own file, so this takes that
  /// file and nothing else.
  Future<void> remove(ManagedCluster cluster) async {
    try {
      await File('${directory.path}/${cluster.sourceFile}').delete();
    } on IOException {
      // Already gone.
    }
    _clusters = _clusters.where((c) => c.id != cluster.id).toList();
    _files = _files.where((f) => f != cluster.sourceFile).toList();
    cluster.dispose();
    if (_selected != null && !_clusters.any((c) => c.id == _selected)) {
      _selected = null;
      _selectSomethingUseful();
    }
    notify();
  }

  /// The other clusters that came out of the same file. They stay.
  List<String> siblings(ManagedCluster cluster) => [
        for (final c in _clusters)
          if (c.sourceFile == cluster.sourceFile && c.id != cluster.id) c.displayName
      ];

  Future<void> probeAll() async {
    await Future.wait([for (final c in _clusters) if (c.isUsable) c.probe()]);
    // Roles are only known once probing has answered, so the default choice
    // is worth revisiting now. The default choice, and only that: a cluster
    // someone chose stays chosen.
    if (_selectionWasAutomatic) {
      _selected = null;
      _selectSomethingUseful();
    }
    notify();
  }

  static final _unsafe = RegExp(r'[^\p{L}\p{N}.-]', unicode: true);

  /// A file name that is one cluster's, and safe on disk.
  static String filenameFor(String context) => '${context.replaceAll(_unsafe, '-')}.kubeconfig';

  /// Anything imported before the split is still one file holding several
  /// clusters. Splitting it on load means an old install behaves like a new
  /// one rather than keeping the coupling for ever.
  Future<bool> _migrateMultiContext(File file) async {
    final List<({String name, String yaml})> parts;
    try {
      parts = Kubeconfig.split(await file.readAsString());
    } catch (_) {
      return false;
    }
    if (parts.length <= 1) return false;
    for (final part in parts) {
      await File('${directory.path}/${filenameFor(part.name)}').writeAsString(part.yaml, flush: true);
    }
    await file.delete();
    return true;
  }

  Future<void> _adopt(File file) async {
    final Kubeconfig config;
    try {
      config = Kubeconfig.parse(await file.readAsString());
    } catch (_) {
      return;
    }
    final name = _basename(file);
    if (!_files.contains(name)) _files.add(name);
    for (final context in config.contexts) {
      if (_clusters.any((c) => c.id == context.name)) continue;
      final cluster = ManagedCluster(context: context, sourceFile: name, words: words);
      cluster.changes.listen((_) => notify());
      _clusters.add(cluster);
    }
  }

  @override
  void dispose() {
    for (final c in _clusters) {
      c.dispose();
    }
    super.dispose();
  }
}
